### ISSUE TITLE
UID SEARCH broken on >= 3.6

### ISSUE THREAD

**Issue by dhke** [cyrusimap/cyrus-imapd#5047]
**Affects:** cyrus-imapd > 3.4 when running in proxy-backend configuration.

The issue appeared when I tried to upgrade our test systems to cyrus-imapd 3.6 from Ubuntu 24.04. The installation almost immediately breaks access e.g. via Thunderbird, which does an `UID SEARCH RECENT` when accessing some mailbox folders. The client gets a 

`Unrecognized UID subcommand`

response from the IMAP server.

I first assumed a packaging or distribution patch problem, but that does not seem to be the case.

If using cyrus-imapd >= 3.6 in proxy mode, the proxy receives

`<76 uid SEARCH UNDELETED SUBJECT "foo"`

but sends to the backend:

`<76 Uid UNDELETED SUBJECT "foo"`

Note the absence of the actual `SEARCH` command on the backend request.

It looks like #4321 (with commit 2e6720278c1e9e0ffdb1504a1d043a1467bf7a2d) broke "UID SEARCH" when using a frontend-backend configuration with imapd acting as proxy. The commit drops `usinguid` from the parameters of `cmd_search()` and does not seem to handle that case before passing the command to the backend:

```
    if (backend_current) {
        /* remote mailbox */
        prot_printf(backend_current->out, "%s %s ", tag, cmd);
        if (!pipe_command(backend_current, 65536)) {
            pipe_including_tag(backend_current, tag, 0);
        }
        return;
    }
```

Since `cmdloop()` consumes the next word after `Uid` using `c = getword(imapd_in, &arg1);`, the actual `SEARCH` command never reaches the backend, which responds with 

`BAD Unrecognized UID subcommand`

The fix might be easy, but it needs to be backported to cyrus 3.6 and 3.8.

Update: Remove suggested patch that breaks esearch.

```
+++ b/imap/imapd.c
@@ -532,7 +532,7 @@ static int parse_fetch_args(const char *tag, const char *cmd,
 static void fetchargs_fini (struct fetchargs *fa);
 static void cmd_fetch(char *tag, char *sequence, int usinguid);
 static void cmd_store(char *tag, char *sequence, int usinguid);
-static void cmd_search(char *tag, char *cmd);
+static void cmd_search(char *tag, char *cmd, int usinguid);
 static void cmd_sort(char *tag, int usinguid);
 static void cmd_thread(char *tag, int usinguid);
 static void cmd_copy(char *tag, char *sequence, char *name, int usinguid, int ismove);
@@ -1811,7 +1811,7 @@ static void cmdloop(void)
             else if (!strcmp(cmd.s, "Esearch")) {
                 if (c != ' ') goto missingargs;
 
-                cmd_search(tag.s, cmd.s);
+                cmd_search(tag.s, cmd.s, 1);
 
                 prometheus_increment(CYRUS_IMAP_ESEARCH_TOTAL);
             }
@@ -2336,8 +2336,7 @@ static void cmdloop(void)
                 if (client_capa & CAPA_UIDONLY) goto uidonly;
                 if (c != ' ') goto missingargs;
             search:
-
-                cmd_search(tag.s, cmd.s);
+                cmd_search(tag.s, cmd.s, usinguid);
 
                 prometheus_increment(CYRUS_IMAP_SEARCH_TOTAL);
             }
@@ -6113,7 +6112,7 @@ static int multisearch_cb(const mbentry_t *mbentry, void *rock)
     return 0;
 }
 
-static void cmd_search(char *tag, char *cmd)
+static void cmd_search(char *tag, char *cmd, int usinguid)
 {
     int c;
     struct searchargs *searchargs;
@@ -6124,23 +6123,24 @@ static void cmd_search(char *tag, char *cmd)
 
     if (backend_current) {
         /* remote mailbox */
-        prot_printf(backend_current->out, "%s %s ", tag, cmd);
+
+        if (usinguid) {
+            /* command was "Uid Search", and "Search" is already consumed */
+            prot_printf(backend_current->out, "%s %s Search ", tag, cmd);
+        } else {
+            prot_printf(backend_current->out, "%s %s ", tag, cmd);
+        }
         if (!pipe_command(backend_current, 65536)) {
             pipe_including_tag(backend_current, tag, 0);
         }
         return;
     }
 
-    switch (cmd[0]) {
-    case 'E':  // Esearch (multisearch)
+    if (cmd[0] == 'E') {
+        // Esearch (multisearch)
         client_behavior.did_multisearch = 1;
         state |= GETSEARCH_SOURCE;
-
-        GCC_FALLTHROUGH
-
-    case 'U':  // Uid Search
-        usinguid = 1;
-        break;
+        // usinguid = 1;
     }
 
     /* RFC 9855, Section 3: MUST reject SEARCH with a charset specification */ 
```

**Comment by jcdelepine** [comment]
Hello,

This bug prevents Gmail mobile clients to join my 3.8.4 murder config. 
It results in an empty mailbox displayed.

On frontend :
<1730042303<9 UID SEARCH 1:* SINCE 26-Oct-2024
>1730042303>9 BAD Unrecognized UID subcommand
<1730042303<10 UID SEARCH 1:* SINCE "26-Oct-2024"
>1730042303>10 BAD Unrecognized UID subcommand
On backend : 
<1730042303<9 Uid 1:* SINCE 26-Oct-2024
>1730042303>9 BAD Unrecognized UID subcommand
<1730042303<10 Uid 1:* SINCE "26-Oct-2024"
>1730042303>10 BAD Unrecognized UID subcommand
<1730042303<Q01 LOGOUT

Shall I apply the patch above or is it this one which "breaks esearch" ?
This patch isn't applied to the 3.10 branch, is there an other one to apply ?

Sincerly,
      Jean Charles Delépine

**Comment by lolorc** [comment]
still running 3.2 here, it looks like this bug has murdered my cyrus upgrade plan :-)
I'm not too sure what the real fix is, doesn't anybody know if this bug also affects 3.10 ? (haven't looked at code yet)

**Comment by dhke** [comment]
If you want to, you can have a go at https://github.com/dhke/cyrus-imapd/tree/fix_proxy_uid_search ([diff](https://github.com/cyrusimap/cyrus-imapd/compare/master...dhke:cyrus-imapd:fix_proxy_uid_search)). 

Note that this is in no way tested, since autoconf currently breaks for me and I currently lack the time to get another dev system up.

**Comment by jcdelepine** [comment]
Thanks a lot, I will give it a try.

I think the last '}' should  be removed in  

```
-    switch (cmd[0]) {
-    case 'E':  // Esearch (multisearch)
+    if (cmd[0] == 'E')
+        // Esearch (multisearch)
+        usinguid = 1
         client_behavior.did_multisearch = 1;
         state |= GETSEARCH_SOURCE;
-
-        GCC_FALLTHROUGH
-
-    case 'U':  // Uid Search
-        usinguid = 1;
-        break;
     }

```

Apart from that the patch backports quite cleanly on 3.8 branch. I just had to ignore ' client_behavior.did_multisearch = 1;' which has been introduce later.

**Comment by jcdelepine** [comment]
Can compile after those smalls typos being corrected : https://github.com/dhke/cyrus-imapd/compare/fix_proxy_uid_search...jcdelepine:cyrus-imapd:patch-1
It's my first github usage, I wish I done it right.

**Comment by jcdelepine** [comment]
UID search works fine on a 3.8.4 murder frontend with the patch.
```
<1731844154<2 UID SEARCH 1:* SINCE 17-Nov-2024
>1731844154>* SEARCH 1643694 1643695 1643696 1643697 1643698 1643699 1643700 1643701 1643702 1643703 1643704 1643705 1643706 1643707 1643708 1643709 1643710
2 OK Completed (17 msgs in 0.003 secs)
<1731844163<. LOGOUT

```

**Comment by dhke** [comment]
Thanks for testing and for the additional fixes. The curly brace needs to go after the `if`, otherwise you always get esearch-flags enabled.

**Comment by jcdelepine** [comment]
I should have seen that the curly brace should be opened, not removed.
UID search still works with the corrected patch. ;-)
Thanks.

**Comment by dhke** [comment]
> Tentative patch for 3.6.6 is here: https://github.com/dhke/cyrus-imapd/tree/fix_proxy_uid_search_3.6.6

I've updated and tested the 3.6.6 version, but I have no idea howto make a pull request for a backport ...

**Comment by ksmurchison** [comment]
Fixed by #5141
